### PR TITLE
Add support for the AWS EFS CSI driver as a k8s add-on

### PR DESCRIPTION
**Pull request by luigidifraia**
### What does this PR do?

Implements support for the EFS CSI driver as a k8s add-on, as requested in issue https://github.com/aws-samples/aws-eks-accelerator-for-terraform/issues/25 previously.

### Motivation

Contribute work back to the community as a token of appreciation for existing work that I use and for those teams I value.

### More

- [x] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [x] Yes, I have added a new example under [examples](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/examples) to support my PR
- [x] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/ssp-eks-add-ons) repo (if applicable)
- [x] Yes, I have updated the [docs](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### Additional Notes

End of the testing scenario (as per example's README.md):

```
kubectl exec efs-app -- bash -c "cat data/out"
Thu Feb 24 00:07:16 UTC 2022
Thu Feb 24 00:07:21 UTC 2022
Thu Feb 24 00:07:26 UTC 2022
Thu Feb 24 00:07:31 UTC 2022
Thu Feb 24 00:07:36 UTC 2022
Thu Feb 24 00:07:41 UTC 2022
Thu Feb 24 00:07:46 UTC 2022
Thu Feb 24 00:07:51 UTC 2022
Thu Feb 24 00:07:56 UTC 2022
```

### CODE REVIEW

File: modules/kubernetes-addons/aws-efs-csi-driver/locals.tf [hunk 1]
```diff
@@ -0,0 +1,55 @@
+locals {
+  service_account_name = "efs-csi-sa"
```
Comment (vara-bonthu) [inline modules/kubernetes-addons/aws-efs-csi-driver/locals.tf:2]: Change the name to 
```hcl
service_account_name = "efs-csi-controller-sa"
```

File: modules/kubernetes-addons/aws-efs-csi-driver/main.tf [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+module "irsa_addon" {
+  source                            = "../../../modules/irsa"
+  create_kubernetes_namespace       = false
+  create_kubernetes_service_account = true
+  eks_cluster_id                    = var.eks_cluster_id
+  kubernetes_namespace              = local.helm_config["namespace"]
+  kubernetes_service_account        = local.service_account_name
+  irsa_iam_policies                 = [aws_iam_policy.aws_efs_csi_driver.arn]
+  tags                              = var.tags
+}
```
Comment (vara-bonthu) [inline modules/kubernetes-addons/aws-efs-csi-driver/main.tf:97]: Remove this module and define locals 

```hcl
locals {
  irsa_config = {
    kubernetes_namespace              = local.namespace
    kubernetes_service_account        = local.service_account_name
    create_kubernetes_namespace       = false
    create_kubernetes_service_account = true
    iam_role_path                     = "/"
    eks_cluster_id                    = var.eks_cluster_id
    irsa_iam_policies                 = concat([aws_iam_policy.aws_efs_csi_driver.arn], var.irsa_policies)
    irsa_iam_permissions_boundary     = var.irsa_iam_permissions_boundary
    tags                              = var.tags
  }
}
```

File: modules/kubernetes-addons/aws-efs-csi-driver/locals.tf [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    max_history                = 0
+    atomic                     = false
+    skip_crds                  = false
+    render_subchart_notes      = true
+    disable_openapi_validation = false
+    dependency_update          = false
+    replace                    = false
+    postrender                 = ""
+  }
+
```
Comment (vara-bonthu) [inline modules/kubernetes-addons/aws-efs-csi-driver/locals.tf:41]: Add this to locals

```hcl
  set_values = [
    {
      name  = "controller.serviceAccount.name"
      value = local.service_account_name
    },
    {
      name  = "controller.serviceAccount.create"
      value = false
    },
    {
      name  = "node.serviceAccount.name"
      value = local.service_account_name
    },
    {
      name  = "node.serviceAccount.create"
      value = false
    }
  ]
```

File: modules/kubernetes-addons/aws-efs-csi-driver/locals.tf (hunk 1 above)
Comment (luigidifraia) [reply modules/kubernetes-addons/aws-efs-csi-driver/locals.tf:2]: Thank you. There are two service account names in recent versions of the AWS EFS CSI driver Helm chart: `efs-csi-controller-sa` and `efs-csi-node-sa`. I tried to create awareness on the fact I am reusing the same SA for both, by generalising its name as `efs-csi-sa`.

File: examples/eks-cluster-with-new-vpc-and-efs/main.tf [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+  enable_aws_efs_csi_driver           = true
+
+  depends_on = [module.aws-eks-accelerator-for-terraform.managed_node_groups]
+}
+
+#--------------
+# Deploy EFS
+#--------------
+resource "aws_efs_file_system" "efs" {
+  creation_token   = "efs"
```
Comment (vara-bonthu) [inline examples/eks-cluster-with-new-vpc-and-efs/main.tf:173]: let's use `encrypted = true` here.

File: modules/kubernetes-addons/aws-efs-csi-driver/locals.tf [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  helm_config = merge(
+    local.default_helm_config,
+    var.helm_config
+  )
+
+  default_helm_values = [templatefile("${path.module}/values.yaml", {
+    service_account_name = local.service_account_name,
+  })]
+
+  set_values = [
```
Comment (luigidifraia) [inline modules/kubernetes-addons/aws-efs-csi-driver/locals.tf:53]: @vara-bonthu At this stage `values.yaml` is not adding any value any longer, but some of the comments therein. Happy to remove `values.yaml` and move comments here, or leave the file around for future defaults to populate `default_helm_values`.

**Comment by luigidifraia** [comment]
Good news. My last [commit](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/pull/266/commits/e6b364d87221eb481d8e93b08bb9597c1e22bf3e) fixed the issue I had when attaching a PV.

**Comment by luigidifraia** [comment]
The issue about EFS encryption reported by `terrascan` seems to be a false positive. This request articulates the reason: https://github.com/accurics/terrascan/issues/801

**Comment by luigidifraia** [comment]
@vara-bonthu @askulkarni2 @kcoleman731 Hope you're keeping well. Is there anything else you'd like me to address for this PR, please?

File: modules/kubernetes-addons/aws-efs-csi-driver/locals.tf [hunk 6]
```diff
@@ -0,0 +1,87 @@
+locals {
+  name                 = "aws-efs-csi-driver"
+  service_account_name = "efs-csi-sa"
+  namespace            = "kube-system"
+
+  default_helm_config = {
```
Comment (kcoleman731) [inline modules/kubernetes-addons/aws-efs-csi-driver/locals.tf:6]: @luigidifraia this is looking great. One request here...please remove values from `default_helm_config` that are default values in the helm_release resource. We are doing clean up on other add-ons this week to do the same.

**Comment by kcoleman731** [comment]
@luigidifraia this is looking great. Thank you for pushing the updates. Left you one last comment, then everything is looking good from my perspective!

File: modules/kubernetes-addons/aws-efs-csi-driver/locals.tf (hunk 6 above)
Comment (luigidifraia) [reply modules/kubernetes-addons/aws-efs-csi-driver/locals.tf:6]: @kcoleman731 Thank you for taking the time and providing feedback. This should now be addressed in my latest commit.

**Review by kcoleman731 — APPROVED** [review]
LGTM!

**Review by askulkarni2 — APPROVED** [review]
Thank you again for your contribution!

File: examples/eks-cluster-with-new-vpc-and-efs/README.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+```shell script
+git clone https://github.com/aws-samples/aws-eks-accelerator-for-terraform.git
+```
+
+#### Step2: Run Terraform INIT
+Initialize a working directory with configuration files
+
+```shell script
+cd examples/eks-cluster-with-new-vpc-and-efs/
```
Comment (vara-bonthu) [inline examples/eks-cluster-with-new-vpc-and-efs/README.md:26]: Could you rename this example folder from `eks-cluster-with-new-vpc-and-efs` to `aws-efs-csi-driver`. Update the README once the renaming is done.

We are planning to rename other example folders as well to easily recognise examples.

**Review by vara-bonthu — APPROVED** [review]
LGTM!

**Pull request merged**